### DISCUSSION TITLE
Some insulin treatments not uploading to nightscout

### DISCUSSION THREAD

**Discussion by jk1993** [NightscoutFoundation/xDrip#3792] [Bug reports]
Hi,

For the last few months I've been using a NovoPen Echo Plus to record mealtime insulin and this has been syncing to nightscout which is great. However in the last couple of weeks some doses have not been uploading to nightscout. If I look in the event log in xdrip I can see the following error which must be related - "Exception uploading REST API treatments":

![Screenshot_20241130_230555_xDrip+](https://github.com/user-attachments/assets/ba707787-e336-4a58-a4b6-9eba7ac60e35)

If I look in the nightscout docker logs I see this, which I think is related as it looks like it's happening in the "save" function in treatments.js:

![docker-error](https://github.com/user-attachments/assets/a4bcad1e-8790-4059-9e5d-b613bcb2a1a3)

Another reason I think it may be related is the thing that seems pretty consistent is that insulin doses under 10 units seem to upload fine, whereas any doses 10 or above seem to get excluded. I'm thinking maybe the value being 10 or above causes the data string being uploaded to nightscout to be too large or use too many characters or something.

Nightscout version: 15.0.2

xDrip version: 2024.11.17

Thanks

**Comment by Navid200** [comment]
Thanks for reporting this.  

@tzachi-dar   
What could be causing this?

**Comment by jk1993** [comment]
Any chance this could be looked into? I'm currently unable to analyse overall insulin usage as any doses of 10 units and above are missing from nightscout

**Reply by dbames** [reply]
Do you have the nightscout logs for the entry the failed to save?  From a quick glance at the nightscout code, it looks to me that it's more likely it is the value of _id that would be the issue rather than the json payload in insulinInjections

**Reply by jk1993** [reply]
The second screenshot is from the nightscout logs, I couldn't see anything else in there at the time it tried to insert the treatment. I guess that makes sense looking at that uuid since it contains the number of units in it, so anything 10.0 or above would have an extra character compared to anything below, assuming this is what it uses to create the id on save.

The issue is still consistent in that all treatments of less than 10 units are fine and anything above 10 units doesn't upload to nightscout. I'm not sure why it's only started happening recently though. Any help would be greatly appreciated and happy to provide more detailed logs

**Comment by jk1993** [comment]
I've done some digging and had a look through my treatments data in nightscout to see exactly when this stopped working. It seems the treatments over 10 units stopped sending to nightscout on the 23rd November 2024. I had a look through my downloads folder on android to see when I last downloaded a new xdrip version and it looks like I downloaded version 20241117 on the 23rd November exactly. It seemed too much of a coincidence to be unrelated but I've tried installing the previous version and I'm still getting the same issue.

Is it possible to convert this post to an issue? I'd really really like to be able to have my insulin treatments syncing to nightscout again as it was incredibly useful to see the data over time.

Thanks

**Reply by Navid200** [reply]
No please.  There is no reason to open an issue.
I just set up a test setup and entered 1 unit of insulin treatment in xDrip and it was uploaded to Nightscout.

Opening an issue does not increase the chance of something being fixed.  But, opening too many issues will drown everything in an ocean of issues.  That will make it worse for me when I need to see what I need to focus on next.

**Comment by alx-k** [comment]
Just wanted to add that I'm having exactly the same problem with exactly the same log output from xDrip and Nightscout.

**Reply by Navid200** [reply]
I need a simple description of the problem that can be recreated.  I give you an example:
Any insulin treatments greater than 10u do not upload to Nightscout.

If you can describe the problem in a short sentence, it is very likely someone will look at it.
I have entered values and they upload to Nightscout.
I need to know how to recreate the problem.

**Reply by alx-k** [reply]
Any insulin treatment entered via NovoPen NFC scan that is over 9u will not upload to Nightscout. TManually entered treatments will upload fine even if they are over 9u.

**Reply by jk1993** [reply]
Thanks for confirming my issue! Just to add (not trying to complicate things) 9.5 units seems to upload fine so its any NovoPen NFC scan where units > 9.5 or units >= 10.0

**Reply by alx-k** [reply]
Good addition, I couldn't test that since my pen cannot do .5 increments.

**Reply by jk1993** [reply]
Makes sense. That also confirms that the issue persists across multiple NovoPen devices, rather than just my model (Echo Plus)

**Reply by Navid200** [reply]
I don't know what is going on with respect to novopen.

Frankly, xDrip does too much.  No single person knows everything that xDrip can do.  That makes it very hard to maintain.  I personally don't record my insulin injections.  So, I am not the right person to fix this.  But, the problem is I am not sure anyone can fix it.  I remember issues with some versions of the pen not working.  I am not sure if we have stopped supporting the pens or not. 

This is still labeled as a bug.  So, it's not that we have forgotten about it.

**Reply by Crosus97** [reply]
> I’m a sysadmin with basic development knowledge, and my partner has diabetes. We rely on Nightscout to sync NovoPen insulin injection data correctly. Since in your last message it seemed you were unsure where to go next, I decided to dig deeper and gather as much relevant information as possible so we can get this fixed.  
>   
> I provided ChatGPT with the following:
> - The `NightscoutUploader.java` file.
> - The `NightscoutTreatments.java` file.
> - Nightscout server logs from my instance.
> - Test results from sending treatments to Nightscout via `curl`.
> - Observations on how xDrip currently formats the `insulinInjections` field and how Nightscout logs it.  
>   
> Based on this, ChatGPT helped me identify the likely cause and propose a patch that should make `insulinInjections` upload as a proper JSON array instead of a string, ensuring correct storage and processing in Nightscout.
>
>PLEASE TAKE THIS AS A GUIDELINE TO HELP NARROW DOWN THE ISSUE, NOT SOMETHING TO BE FOLLOWED TO THE LETTER.

**Summary**  
When xDrip uploads NovoPen treatments, the `insulinInjections` field is currently sent as a JSON *string* instead of a real JSON array. This makes Nightscout store it as plain text and can cause downstream processing issues. Sending a proper array fixes it.

**Tests done**  
Against my Nightscout instance:
- With **string** → created, but stored as plain text.
- With **array** → created and parsed as an array; works as expected.

**Root cause**  
In `NightscoutUploader.java`, the uploader places `treatment.insulinJSON` (a serialized JSON string) directly into the payload, instead of sending an actual array.

**Proposed fix (xDrip upload)**  
Parse `treatment.insulinJSON` into a `JSONArray` and put that into the payload (so Nightscout receives a real array).

```diff
diff --git a/app/src/main/java/com/eveningoutpost/dexdrip/utilitymodels/NightscoutUploader.java b/app/src/main/java/com/eveningoutpost/dexdrip/utilitymodels/NightscoutUploader.java
@@ -XXXX,7 +XXXX,17 @@ private void populateV1APITreatmentEntry(org.json.JSONArray array, Treatments treatment) throws Exception {
-        if (treatment.insulinJSON != null) {
-            record.put("insulinInjections", treatment.insulinJSON);
-        }
+        if (treatment.insulinJSON != null) {
+            // Ensure insulinInjections is a real JSON array (not a quoted JSON string)
+            try {
+                record.put("insulinInjections", new org.json.JSONArray(treatment.insulinJSON));
+            } catch (org.json.JSONException e) {
+                // If it isn't valid JSON, omit the field rather than sending malformed data
+                // (or optionally wrap as an empty array)
+                // record.put("insulinInjections", new org.json.JSONArray());
+            }
+        }
```

**Optional companion fix (download)**  
Make `NightscoutTreatments.java` tolerant of both array and string so round-trip remains consistent.

```diff
diff --git a/app/src/main/java/com/eveningoutpost/dexdrip/utilitymodels/NightscoutTreatments.java b/app/src/main/java/com/eveningoutpost/dexdrip/utilitymodels/NightscoutTreatments.java
@@ -XXXX,7 +XXXX,19 @@ public class NightscoutTreatments {
-            try {
-                injections = tr.getString("insulinInjections");
-            } catch (org.json.JSONException e) {
-                // ignore
-            }
+            try {
+                // Accept either a JSON array or a string
+                Object inj = tr.get("insulinInjections");
+                if (inj instanceof org.json.JSONArray) {
+                    injections = ((org.json.JSONArray) inj).toString();
+                } else if (inj instanceof String) {
+                    injections = (String) inj;
+                }
+            } catch (org.json.JSONException e) {
+                // ignore
+            }
```

**Expected outcome**  
Nightscout receives `insulinInjections` as a real array, stores it as structured data, and downstream processing works correctly for NovoPen treatments.


---

*I apologize if anyone feels uncomfortable or offended by the fact that I tried to resolve this issue with ChatGPT-5 (Plus user). My only intention is to help get this fixed so that NovoPen insulin data syncs properly for those who need it.*

**Comment by jk1993** [comment]
To provide an update to this, my Novopen Echo Plus stopped working altogether. The doses wouldn't even record on the device itself and the on-device screen just read "Error". So since it wasn't recording doses at all I got a replacement pen. Interestingly enough, with the replacement pen, doses of 10+ units uploaded to nightscout just fine.

This was back in June, and since it was now uploading all doses to nightscout, I haven't had an issue. Fast forward to a couple of days ago, the new pen has stopped uploading doses of 10+ units to nightscout. Interestingly, it's been about 4 months since I switched over pens (June - Oct), and about 4 months with the previous pen the first time this happened (July - November 2024). So that probably makes it even more difficult to replicate the issue, since it seems as though it's a combination of the issue, as well as the age of the pen.

Not sure if this will help but thought I'd share my experience and provide as much information as possible. It does mean there is a temporary workaround I suppose (get a new pen!)

**Comment by vytux-com** [comment]
I think I have found the issue and the solution, just need to do some more local tests on a cleaner build without any other changes.

<img width="503" height="238" alt="image" src="https://github.com/user-attachments/assets/482b18d5-fa3e-413b-8369-42761d4d19e3" />

but I believe this change will fix this as the issue appears to be that the `uuid_to_id` function assumes that any uuid with length of 24 is already valid, whereas the uuid of the NFC scan includes the dose and if the dose is more than 10 units it just so happens to be 24 chars long. 

e.g. 
```
963155e309f64404840511fc
ABIXSJ:Open67985740:38.0
```

**Reply by Navid200** [reply]
@vytux-com Fantastic!

Please let me know if you need help after you are done with your tests.

**Reply by vytux-com** [reply]
Done and tests successful with no other changes just that 1 if condition.

What is the best way to process this PR? Just create one based on the main branch? Some other branch etc?

**Reply by vytux-com** [reply]
https://github.com/NightscoutFoundation/xDrip/pull/4332

**Reply by vytux-com** [reply]
@Navid200 do you know if there is a easy (ish) way to sign the custom build apk?  My main phone is blocked from installing APK that cannot be verified by Play Protect. So although I've been able to test on my spare phone I will need to wait for a while to install an official build unless there is a way to self sign?

**Reply by Navid200** [reply]
I don't follow.  What is special about your main phone.

You must have your own key to compile.  Don't you?
Or did you only test this with a virtual device?

I don't want to go too much into detail answering the wrong question.
Can you clarify what you are asking?

**Reply by Navid200** [reply]
You cannot install your compiled version over the official installation because the two use different keys.
Is that the problem?

**Reply by Navid200** [reply]
You can first create a backup: https://navid200.github.io/xDrip/docs/Backup-Database.html 
Considering you have Nightscout, you have all your data backed up on Nightscout as well.
Now, after backup, you can uninstall the official release.   You can then install your compiled version.  
It will not be able to run Google Drive backup.  But, I believe it can do everything else.  That is what I use.  
If you ever want to go  back to the official release, you have have to do the same thing again meaning uninstall first and then install.

**Reply by vytux-com** [reply]
OK, it most likely is the keys then. I also have a 'work' profile setup and that has enabled some features that cannot be turned off. 

As for testing, it was done on a full physical device, just my spare phone.

**Reply by vytux-com** [reply]
On my main phone I just tried the uninstall & install the self-compiled version but since the app needs "sensitive" permissions it's being blocked by play protect and I cannot disable that (even temporarily) due to the work profile security policy. Reinstalling the official release (2025-09-05 version) came up with the same warning but it had an option to install anyway.

**Reply by Navid200** [reply]
When I install my compiled version, I see the same behavior as the official release.  It shows that the app is for a previous version and there is the install anyway option.  https://navid200.github.io/xDrip/docs/FAQ/UnsafeAppBlocked.html

**Reply by vytux-com** [reply]
When installing my version I get this popup
<img width="1080" height="2340" alt="Screenshot_20260109-141052" src="https://github.com/user-attachments/assets/b26c1f05-0502-4839-9496-3385d266bc2b" />

When doing downloaded version I get this
<img width="1080" height="2340" alt="Screenshot_20260109-141119" src="https://github.com/user-attachments/assets/10b2b6cc-0ab5-4a36-b731-503c10c24821" />

To build the apk I just used the Android studio, got clone the 2025.09.05 tag, change the 1 file, then Build -> Build Bundles -> Build APKs 

At no stage did it ask for any signing keys, so it either generated some automatically or used none. When build is done I copy the debug/app-fast-debug.apk to the phone and install.

**Reply by Navid200** [reply]
I see.  I will need to add more detail to our guide.  Thanks.  

To compile, do the following in sequence.  After each step, wait for all processes to complete before moving to the next step.  
If you have never compiled the project, you can skip the first 2 steps.

Build > Clean project
Build > Rebuild project
Build > Generate signed bundle / APK 

After starting the third step, choose:
APK

The very first time you get to this point using Android Studio, you need to choose:
Create new

Choose a path on your computer.  Choose an alias you will remember.
Choose a password you will remember or take a note.

After that, every time you get to this point, you choose:
Choose existing
Then, enter find the key on your computer, enter the alias and password.

Choose:
ProdRelease

Let me know if you have any questions.

**Reply by vytux-com** [reply]
I just did all of that, however I am still getting the same screen without the option to Install Anyway

**Reply by Navid200** [reply]
If you attempt to install it on your other phone, do you get the option to install anyway?

**Reply by Navid200** [reply]
Congratulations!  Your PR was merged.

Next time you are compiling, if you need to do this again, the output will be in Project/app/prod/release

**Reply by vytux-com** [reply]
I have installed the 2025-01-09 nightly and can confirm it now uploads all the values, and if you do backfill it will even upload the old historical values that were missed before

**Reply by Navid200** [reply]
Great!
Thanks a lot

**Reply by jk1993** [reply]
Can confirm it's all working perfectly for me after updating to the latest nightly. Thanks so much for fixing this!
